File: genium/src/main/java/com/here/genium/model/java/JavaClass.java
/*
 * Copyright (C) 2016-2018 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.model.java;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Stream;

public final class JavaClass extends JavaTopLevelElement {

  public final Set<JavaField> fields = new LinkedHashSet<>();
  public final Set<JavaField> parentFields = new LinkedHashSet<>();
  public final JavaType extendedClass;
  public final boolean isImplClass;
  public final boolean needsDisposer;
  public final boolean isParcelable;
  public final boolean isEquatable;

  public JavaClass(final String name) {
    this(name, null, false, false, false, false);
  }

  @lombok.Builder(builderClassName = "Builder")
  private JavaClass(
      final String name,
      final JavaType extendedClass,
      final boolean isImplClass,
      final boolean needsDisposer,
      final boolean isParcelable,
      final boolean isEquatable) {
    super(name);
    this.extendedClass = extendedClass;
    this.isImplClass = isImplClass;
    this.needsDisposer = needsDisposer;
    this.isParcelable = isParcelable;
    this.isEquatable = isEquatable;
  }

  @SuppressWarnings("unused")
  public boolean needsBuilder() {
    return parentFields.size() + fields.size() > 2;
  }

  @SuppressWarnings("unused")
  public Collection<JavaField> getAllVisibleFields() {
    List<JavaField> result = new LinkedList<>();
    result.addAll(parentFields);
    result.addAll(fields);
    return result;
  }

  @Override
  public Stream<JavaElement> stream() {
    Stream<? extends JavaElement> extendedClassStream =
        extendedClass != null ? extendedClass.stream() : Stream.empty();
    return Stream.of(super.stream(), fields.stream(), extendedClassStream)
        .flatMap(Function.identity());
  }

  @Override
  public Set<JavaImport> getImports() {

    Set<JavaImport> imports = super.getImports();

    if (extendedClass != null) {
      imports.addAll(extendedClass.imports);
      // No need to import things from the same package. This also filters out a self-import.
      imports.removeIf(anImport -> Objects.equals(anImport.javaPackage, this.javaPackage));
    }

    return imports;
  }

  public static Builder builder(final String name) {
    return new Builder().name(name);
  }
}

File: genium/src/main/java/com/here/genium/model/java/JavaClass.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.model.java

class JavaClass(
    name: String,
    val extendedClass: JavaType? = null,
    val fields: List<JavaField> = emptyList(),
    methods: List<JavaMethod> = emptyList(),
    constants: List<JavaConstant> = emptyList(),
    val isImplClass: Boolean = false,
    val needsDisposer: Boolean = false,
    val isParcelable: Boolean = false,
    val isEquatable: Boolean = false,
    val hasNativeEquatable: Boolean = false,
    val isImmutable: Boolean = false,
    val needsBuilder: Boolean = false,
    var generatedConstructorComment: String? = null
) : JavaTopLevelElement(name) {

    init {
        this.methods += methods
        this.constants += constants
    }

    @Suppress("unused")
    val uninitializedFields = fields.filter { !it.initial.isCustom }

    @Suppress("unused")
    val defaultedFields = fields.filter { it.initial.isCustom }

    @Suppress("unused")
    val hasDefaults = uninitializedFields.size < fields.size

    @Suppress("unused")
    val allArgsConstructorVisibility: JavaVisibility
        get() = when {
            fields.any { it.visibility == JavaVisibility.PACKAGE } -> JavaVisibility.PACKAGE
            else -> visibility
        }

    @Suppress("unused")
    val freeArgsConstructorVisibility: JavaVisibility
        get() = when {
            uninitializedFields.any { it.visibility == JavaVisibility.PACKAGE } ->
                JavaVisibility.PACKAGE
            else -> visibility
        }

    @Suppress("unused")
    val constructors
        get() = methods.filter { it.isConstructor }

    override fun stream(): List<JavaElement> {
        val extendedClassStream = extendedClass?.stream() ?: emptyList()
        val implClassStream =
            when {
                isParcelable || needsBuilder ->
                    fields
                        .map { it.type }
                        .filterIsInstance<JavaTemplateType>()
                        .map { it.implementationType }
                else -> emptyList()
            }
        return super.stream() + fields + extendedClassStream + implClassStream
    }

    override val imports
        get() =
            if (extendedClass != null) {
                val parentImports =
                    extendedClass.imports.filterNot { it.javaPackage == this.javaPackage }
                (super.imports + parentImports).toSortedSet()
            } else {
                super.imports
            }
}
